provenance_lib/parse.py

```diff
@@ -0,0 +1,362 @@
+        """ depth-first traversal of this ProvNode's ancestors """
+        local_parents = dict()
+        if not self.parents:
+            local_parents = {self.uuid: None}
+        else:
+            sub_dag = dict()
+            for parent in self.parents:
+                sub_dag.update(parent.traverse_uuids())
+            local_parents[self.uuid] = sub_dag
+        return local_parents
+
+
+class Archive:
+    """
+    Lightly-processed contents of a single QIIME 2 Archive, an Archive
+    contains a non-hierarchical pool of unique ProvNodes
+    """
+
+    # TODO: Can we assume all Q2 archives contain only unique Results?
+    # If not, Archive should filter duplicates.
+
+    # TODO: add property for archive version?
+    # TODO: UUID class with basic validation?
+    # NOTE: static type declarations like this break testing unless the class
+    # is forward declared (ProvNode must be above Archive or NameError).
+    _number_of_results: int
+    _archive_contents: Dict[str, ProvNode]
+    _archive_md: _ResultMetadata
+
+    @property
+    def root_uuid(self):
+        return self._archive_md.uuid
+
+    @property
+    def archive_type(self):
+        return self._archive_md.type
+
+    @property
+    def archive_format(self):
+        return self._archive_md.format
+
+    def get_result(self, uuid):
+        return self._archive_contents[uuid]
+
+    def __str__(self):
+        return f"Archive(Root: {self.root_uuid})"
+
+    def __repr__(self):
+        r_str = (f"Archive(Root: {self.root_uuid}, Semantic Type: "
+                 f"{self.archive_type}, Format: {self.archive_format})\n"
+                 "Contains Results:\n")
+        for uuid in self._archive_contents.keys():
+            r_str += str(uuid)
+            r_str += "\n"
+        return r_str
+
+    def __init__(self, archive_fp: str):
+        self._archive_md: None
+        self._archive_contents: None
+        self._number_of_results = 0
+
+        # Get archive metadata including root uuid
+        with zipfile.ZipFile(archive_fp) as zf:
+            all_filenames = zf.namelist()
+            root_metadata_fps = filter(self._is_root_metadata_file,
+                                       all_filenames)
+
+            try:
+                root_metadata_fp = next(root_metadata_fps)
+            except StopIteration:
+                raise ValueError("Malformed Archive: "
+                                 "no top-level metadata.yaml file")
+
+            try:
+                next(root_metadata_fps)
+                raise ValueError("Malformed Archive: "
+                                 "multiple top-level metadata.yaml files")
+            except StopIteration:
+                pass
+
+            self._archive_md = _ResultMetadata(zf, root_metadata_fp)
+            self._populate_archive(zf)
+
+    def _populate_archive(self, zf: zipfile):
+        self._archive_contents = {}
+
+        prov_data_filenames = filter(self._is_prov_data, zf.namelist())
```

Comment:
I recommend globbing for these rather than doing the filtering, since for the most part you know what files you're looking for. That should also let you get the uuids as well.

Reply:
As above, I'd _love_ to go this route, but I don't think I can glob inside of a zipfile.

Reply:
edc2e9a removes most of the filtering, and tidies this function up a lot